src/main/scala/firrtl/transforms/TopWiring.scala

```diff
@@ -0,0 +1,247 @@
+// See LICENSE for license details.
+package firrtl.transform
+package TopWiring
+
+import firrtl._
+import firrtl.ir._
+import firrtl.passes._
+import firrtl.annotations._
+import firrtl.Mappers._
+import firrtl.graph._
+
+import java.io._
+import scala.io.Source
+import collection.mutable
+
+import firrtl.passes.wiring._
+
+/** Annotation for optional output files **/
+case class TopWiringOutputFilesAnnotation(dirname: String, outputfunction: (String,String,Seq[((ComponentName, Type, Boolean, Seq[String]), Int)], CircuitState) => CircuitState) extends NoTargetAnnotation
```

Comment:
I'd be inclined to use a case class for the inputs to the `outputfunction` for the sake of clarity (also, I think `dirName` and `outputFunction` for Scala naming conventions).